src/braket/default_simulator/simulation_strategies/batch_operation_strategy.py

```diff
@@ -44,46 +46,90 @@ def apply_operations(
         (num_qubits, 0) tensor
     """
-    # TODO: Write algorithm to determine partition size based on operations and qubit count
-    partitions = [operations[i : i + batch_size] for i in range(0, len(operations), batch_size)]
+    if not operations:
+        return state
+
+    if batch_size == 1:
+        return _apply_operations_sequential(state, operations)
 
-    for partition in partitions:
-        state = _contract_operations(state, qubit_count, partition)
+    return _apply_operations_batched(state, qubit_count, operations, batch_size)
 
+
+def _apply_operations_sequential(state: np.ndarray, operations: list[GateOperation]) -> np.ndarray:
+    """Apply operations sequentially without batching."""
+    for op in operations:
+        state = _apply_operation(state, op)
     return state
 
 
-def _contract_operations(
-    state: np.ndarray, qubit_count: int, operations: list[GateOperation]
+def _apply_operations_batched(
+    state: np.ndarray, qubit_count: int, operations: list[GateOperation], batch_size: int
 ) -> np.ndarray:
-    contraction_parameters = [state, list(range(qubit_count))]
+    """Apply operations in optimized batches."""
+    current_batch = []
+
+    def process_batch():
+        nonlocal state
+        if current_batch:
+            state = _process_optimized_batch(state, qubit_count, current_batch)
+            current_batch.clear()
+
+    for op in operations:
+        current_batch.append(op)
+        if len(current_batch) >= batch_size:
+            process_batch()
+
+    process_batch()
+    return state
+
+
+def _apply_operation(state: np.ndarray, op: GateOperation):
+    """Apply an operation to the state."""
+    matrix = op.matrix
+    all_targets = op.targets
+    num_ctrl = len(op._ctrl_modifiers)
+    control_state = op._ctrl_modifiers
+    controls = all_targets[:num_ctrl]
+    targets = all_targets[num_ctrl:]
+    return multiply_matrix(state, matrix, targets, controls, control_state)
+
+
+def _process_optimized_batch(state, qubit_count, operations):
+    """Process a batch of operations with optimized tensor contraction."""
+    if len(operations) <= 2:
+        for op in operations:
+            state = multiply_matrix(state, op.matrix, op.targets, [], [])
+        return state
+
+    contraction_parameters = [state, [*range(qubit_count)]]
     index_substitutions = {i: i for i in range(qubit_count)}
     next_index = qubit_count
+
     for operation in operations:
         matrix = operation.matrix
         targets = operation.targets
 
-        # Lower indices, which will be traced out
         covariant = [index_substitutions[i] for i in targets]
+        contravariant = [*range(next_index, next_index + len(targets))]
+        indices = contravariant + covariant
 
-        # Upper indices, which will replace the contracted indices in the state vector
-        contravariant = list(range(next_index, next_index + len(covariant)))
+        matrix_dim = int(np.log2(matrix.shape[0]))
```

Comment:
Is this necessary? `matrix_dim * 2` is already equal to `len(contravariant + covariant)`